File: C++/HackerRank_CPP/HackerRank_CPP/NonDivisibleSubset.cpp
//
//  NonDivisibleSubset.cpp
//  HackerRank_CPP
//
//  Created by Sebastian Tleye on 02/09/2025.
//

#include "NonDivisibleSubset.hpp"

void NonDivisibleSubset::solve() {
    //nonDivisibleSubset(4, {19, 10, 12, 10, 24, 25, 22});
    nonDivisibleSubset(1, {1});
}

int NonDivisibleSubset::nonDivisibleSubset(int k, vector<int> s) {

    int res {};
    vector<int> remainders(k, 0);

    for (auto e: s) {
        remainders.at(e % k) += 1;
    }
    
    if ((remainders.at(0)) > 0) {
        res++;
    }

    for (int i = 1; i < ceil(k/2.0); i++) {
        res += max(remainders.at(i), remainders.at(k-i));
    }
    
    if (k % 2 == 0) {
        if (remainders.at(k/2) > 0) {
            res += 1;
        }
    }

    return res;

}

File: Python/non_divisible_subset.py
import math


def nonDivisibleSubset(k, s):
    res = 0

    remainders = [0] * k

    for e in s:
        remainders[e % k] += 1

    if remainders[0] > 0:
        res += 1

    for i in range(1, math.ceil(k/2.0)):
        res += max(remainders[i], remainders[k-i])

    if k % 2 == 0:
        if remainders[k//2] > 0:
            res += 1

    return res


# print(nonDivisibleSubset(4, [19, 10, 12, 10, 24, 25, 22]))
print(nonDivisibleSubset(3, [1, 7, 2, 4]))
